### PR TITLE
[WIP] introduce a region unification table and use it in dropck

### LINKED ISSUE #29844: Regression: overflow while adding drop-check rules
https://users.rust-lang.org/t/drop-check-regression-overflow-while-adding-drop-check-rules-for-struct/3617

```rust
use std::sync::Arc;

pub struct DescriptorSet<'a> {
    pub slots: Vec<AttachInfo<'a, Resources>>
}

pub trait ResourcesTrait<'r>: Sized {
    type DescriptorSet: 'r;
}

pub struct Resources;

impl<'a> ResourcesTrait<'a> for Resources {
    type DescriptorSet = DescriptorSet<'a>;
}

pub enum AttachInfo<'a, R: ResourcesTrait<'a>> {
    NextDescriptorSet(Arc<R::DescriptorSet>)
}

fn main() {
    let _x = DescriptorSet {slots: Vec::new()};
}
```

[playpen](http://is.gd/UarpVI)

gives the error

```text


<anon>:22:9: 22:11 error: overflow while adding drop-check rules for DescriptorSet<'_> [E0320]
<anon>:22     let _x = DescriptorSet {slots: Vec::new()};
                  ^~
<anon>:22:9: 22:11 note: overflowed on enum AttachInfo variant NextDescriptorSet field #0 type: alloc::arc::Arc<DescriptorSet<'_>>
<anon>:22     let _x = DescriptorSet {slots: Vec::new()};
                  ^~
<anon>:22:14: 22:47 error: overflow while adding drop-check rules for DescriptorSet<'_> [E0320]
<anon>:22     let _x = DescriptorSet {slots: Vec::new()};
                       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
<anon>:22:14: 22:47 note: overflowed on enum AttachInfo variant NextDescriptorSet field #0 type: alloc::arc::Arc<DescriptorSet<'_>>
<anon>:22     let _x = DescriptorSet {slots: Vec::new()};
                       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
<anon>:22:36: 22:46 error: overflow while adding drop-check rules for collections::vec::Vec<AttachInfo<'_, Resources>> [E0320]
<anon>:22     let _x = DescriptorSet {slots: Vec::new()};
                                             ^~~~~~~~~~
<anon>:22:36: 22:46 note: overflowed on struct alloc::arc::Arc field `_ptr` type: core::ptr::Shared<alloc::arc::ArcInner<DescriptorSet<'_>>>
<anon>:22     let _x = DescriptorSet {slots: Vec::new()};
                                             ^~~~~~~~~~
error: aborting due to 3 previous error
```

cc @pnkfelix

### PR DESCRIPTION
**Pull request by arielb1**
Fixes #29844 

I would prefer to
(a) make some performance measurements
(b) use the unification table in a few more places
before committing further, but this is probably good enough for beta.

r? @nikomatsakis

### CODE REVIEW

File: src/librustc/middle/infer/resolve.rs [hunk 1]
```diff
@@ -45,6 +45,41 @@ impl<'a, 'tcx> ty::fold::TypeFolder<'tcx> for OpportunisticTypeResolver<'a, 'tcx
     }
 }
 
+/// The opportunistic type and region resolver is similar to the
+/// opportunistic type resolver, but also opportunistly resolves
+/// regions. It is useful for canonicalization.
```
Comment (nikomatsakis) [inline src/librustc/middle/infer/resolve.rs]: Why not just always resolve regions?

File: src/librustc/middle/infer/region_inference/mod.rs [hunk 2]
```diff
@@ -460,6 +468,10 @@ impl<'a, 'tcx> RegionVarBindings<'a, 'tcx> {
             // equating regions.
             self.make_subregion(origin.clone(), sub, sup);
             self.make_subregion(origin, sup, sub);
+
+            if let (ty::ReVar(sub), ty::ReVar(sup)) = (sub, sup) {
+                self.unification_table.borrow_mut().union(sub, sup);
+            }
```
Comment (nikomatsakis) [inline src/librustc/middle/infer/region_inference/mod.rs]: Hmm. I expected to not add two subregion edges here, but rather to "refresh" the regions in the inference propagation below.

**Comment by nikomatsakis** [comment]
OK, so, re-reading your comment, I expect that you were probably shooting for a good way to make a kind of minimally invasive (and minimally risky) version of this PR, which makes a lot of sense.

**Comment by nikomatsakis** [comment]
I like this better than the alternative I proposed, and my main complaints about this PR was that it was not going as far as I would expect -- but that's probably good from the POV of being minimally invasive. 
So:

@bors r+

**Comment by arielb1** [comment]
@nikomatsakis 

This commit causes a 50% regression in typeck performance (not the WIP). I will write a separate patch to fix it.

**Comment by pnkfelix** [comment]
> I will write a separate patch to fix it.

Namely PR #30368 (right?)

**Comment by arielb1** [comment]
Right that was #30368

**Pull request merged**